### PR TITLE
Patient list for saved cohorts

### PR DESCRIPTION
**Pull request by nfreundl**
Implementation of patient list retrieval from saved cohorts, the authorization as patient_list for Explore queries.

### CODE REVIEW

File: test/test_cli_preloading.sh [hunk 1]
```diff
@@ -346,6 +404,10 @@ test7 "samevisit" "samevisit" "samevisit" "samevisit" "${timingResultZeroExpecte
 test7 "sameinstancenum" "sameinstancenum" "any" "sameinstancenum" "${timingResultNonZeroExpected}"
 test7 "samevisit" "samevisit" "any" "samevisit" "${timingResultNonZeroExpected}"
 
+echo "Testing cohorts-patient-list"
+USERNAME=${1:-test}
```
Comment (f-marino) [inline test/test_cli_preloading.sh]: You are not using USERNAME inside test8, why redefining it here?

File: connector/client/querytools/cli.go [hunk 2]
```diff
@@ -196,3 +199,72 @@ func ExecuteRemoveCohorts(token, username, password, cohortName string, disableT
...
+	}
+	patientLists, nodeTimers, localTimers, err := cohortsPatientList.Execute()
+	if err != nil {
+		err = fmt.Errorf("while executing patient list request, request ID %s: %s", cohortsPatientList.id, err.Error())
+		logrus.Error(err.Error())
+		return err
+	}
+
+	// displaying results
+	resultCSV, err := utilclient.NewCSV(resultFile)
```
Comment (f-marino) [inline connector/client/querytools/cli.go:235]: resultFile may not be defined (empty string), this should not be considered an error, in this case you should just print the result to stdout.

File: connector/client/querytools/cli.go [hunk 3]
```diff
@@ -196,3 +199,72 @@ func ExecuteRemoveCohorts(token, username, password, cohortName string, disableT
...
+	}
+	err = resultCSV.Close()
+	if err != nil {
+		err = fmt.Errorf("cohorts patient list request closing result file: %s", err.Error())
+		logrus.Error(err)
+		return err
+	}
+
+	// dumping timers
+	err = medcoclient.DumpTimers(timerFile, nodeTimers, localTimers)
```
Comment (f-marino) [inline connector/client/querytools/cli.go:267]: do you think it is useful to define a different file for the timers? Couldn't resultFile be used? I let you judge on this.

File: cmd/medco-cli-client/main.go [hunk 4]
```diff
@@ -163,6 +163,20 @@ func main() {
...
+	//--- query tools command flags
+	cohortsPatientListFlag := []cli.Flag{
+		cli.StringFlag{
+			Name:     "cohortName, c",
+			Usage:    "Name of the new cohort",
+			Required: true,
+		},
+		cli.StringFlag{
+			Name:     "dumpFile, d",
+			Usage:    "File for dumping ",
```
Comment (f-marino) [inline cmd/medco-cli-client/main.go]: this flag needs a better description

File: cmd/medco-cli-client/main.go [hunk 5]
```diff
@@ -395,6 +409,25 @@ func main() {
 				)
 			},
 		},
+		{
+			Name:        "cohorts-patient-list",
+			Aliases:     []string{"cpl"},
+			Usage:       "Request the patients' numbers.",
+			Flags:       cohortsPatientListFlag,
+			ArgsUsage:   "-c cohortName [-d timer dump file]",
+			Description: "Retrieves the numbers of the patients associated to a cohort.",
```
Comment (f-marino) [inline cmd/medco-cli-client/main.go]: A bit confusing this Usage and Description. Why not just mirroring the description of the backend API you are calling?

File: connector/server/querytools/saved_cohorts.go [hunk 6]
```diff
@@ -199,6 +221,16 @@ func RemoveCohort(userID, cohortName string) error {
 	return nil
 }
```
Comment (f-marino) [inline connector/server/querytools/saved_cohorts.go]: @mickmis do we allow to perform queries like that? Aren't all queries supposed to be implemented as stored procedures?

**Review by f-marino — CHANGES_REQUESTED** [review]

File: connector/client/querytools/cli.go (hunk 2 above)
Comment (nfreundl) [reply connector/client/querytools/cli.go:235]: if the strings is empty, utilclient.NewCSV does not throw an error and return the stdout file descriptor, doing what you suggest

File: connector/client/querytools/cli.go (hunk 3 above)
Comment (nfreundl) [reply connector/client/querytools/cli.go:267]: I'd rather write different files for the results and other informations about ran procedures, but resultFile could be used.

I don't have a strong opinion about that.

File: connector/server/querytools/saved_cohorts.go (hunk 6 above)
Comment (mickmis) [reply connector/server/querytools/saved_cohorts.go]: @f-marino you are right, queries should be stored procedures for compliance with the HUG security requirements.

File: build/package/i2b2/sql/85-data-manipulation-functions/table_name.plpgsql [hunk 7]
```diff
@@ -0,0 +1,16 @@
+-- pl/pgsql function that returns the table name for a given table code
```
Comment (mickmis) [inline build/package/i2b2/sql/85-data-manipulation-functions/table_name.plpgsql]: You can move this file to `30-stored-procedures`. This folder is for pl/pgsql functions that are used manually for some operations, not for procedures used in the medco runtime.

File: connector/server/querytools/cohorts_patient_list_logic.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+	"fmt"
+	"time"
+
+	medcomodels "github.com/ldsec/medco/connector/models"
+	"github.com/ldsec/medco/connector/restapi/models"
+	"github.com/ldsec/medco/connector/wrappers/unlynx"
+	"github.com/sirupsen/logrus"
+)
+
+// CohortsPatientList holds the parameters structure to retriev the patient list of a saved cohort
```
Comment (mickmis) [inline connector/server/querytools/cohorts_patient_list_logic.go]: retrieve typo

**Review by mickmis — CHANGES_REQUESTED** [review]

File: build/package/i2b2/sql/85-data-manipulation-functions/table_name.plpgsql (hunk 7 above)
Comment (nfreundl) [reply build/package/i2b2/sql/85-data-manipulation-functions/table_name.plpgsql]: functions in stored procedures are bound to tables created within by medco package.

The functions you mentionned interacted with i2b2 tables. May I put them in a new stored-procedure folder for i2b2 build package as well ?

Comment (mickmis) [reply build/package/i2b2/sql/85-data-manipulation-functions/table_name.plpgsql]: OK I see, then yep you should put that into a new folder.

Comment (nfreundl) [reply build/package/i2b2/sql/85-data-manipulation-functions/table_name.plpgsql]: This is done in cb4ff05ccfb74cf83e34bc760d7756d3c9d2b9f6, can you check if this is fine ?

**Review by mickmis — APPROVED** [review]

**Review by f-marino — APPROVED** [review]

**Pull request merged**